src/services/domainService.ts

```diff
@@ -1,177 +1,44 @@
-  serviceFn: Function,
-  context: Context,
-  request: HttpRequest
-) {
-  let dbClient: Maybe<MongoClient>;
-
-  try {
-    dbClient = await getConnectedDbClient(context);
-    const dbService: MongoDbService = getDatabaseService(dbClient);
-    await serviceFn(context, request, dbService);
-  } catch (e: any) {
-    const error: Error = {
-      name: e.name ?? "Unknown Internal Server Error",
-      message: e.message ?? "An unexpected server error occured.",
-    };
-    context.log(e.name, e.message, e.stack);
-
-    context.res = {
-      status: 500,
-      body: `${error.name} \n ${error.message}`,
-    };
-  } finally {
-    if (dbClient) {
-      await dbClient.close();
-    }
-  }
-}
-
-export async function listDomains(
-  context: Context,
-  request: HttpRequest,
-  dbService: MongoDbService
-) {
-  let findOptions: Maybe<DomainFindOptions>;
-  if (request.body) {
-    findOptions = request.body.options;
-  }
-  const results = await dbService.getAllDomains(findOptions);
-
-  const responseBody = {
-    numResults: results.length,
-    results: results,
-  };
-
-  context.res = {
-    body: responseBody,
-  };
-}
-
-export async function getDomain(
-  context: Context,
-  request: HttpRequest,
-  dbService: MongoDbService
-) {
-  const domainId: DomainId = request.params.id;
-  if (!validateDomainId(domainId)) {
-    context.res = {
-      status: 400,
-      body: validateDomainId.errors,
-    };
-    return;
-  }
-
-  let findOptions: Maybe<DomainFindOptions>;
-  if (request.body) {
-    findOptions = request.body.options;
-  }
-
-  const result = await dbService.getValidDomain(request.params.id, findOptions);
-
-  const responseBody = {
-    result: result,
-  };
-  context.res = {
-    body: responseBody,
-  };
-}
-
-export async function getSubdomains(
-  context: Context,
-  request: HttpRequest,
-  dbService: MongoDbService
-) {
-  const domainId: DomainId = request.params.id;
-  if (!validateDomainId(domainId)) {
-    context.res = {
-      status: 400,
-      body: validateDomainId.errors,
+import { Domain, DomainFindOptions, Maybe } from "@zero-tech/data-store-core";
+import { DomainResponse, DomainsResponse } from "../types";
+
+export abstract class DomainService<T> {
+  dbService: T;
+  constructor(service: T) {
+    this.dbService = service;
+  }
+  abstract listDomains(
+    findOptions: Maybe<DomainFindOptions>
+  ): Promise<DomainsResponse>;
+  abstract getDomain(
+    id: string,
+    findOptions: Maybe<DomainFindOptions>
+  ): Promise<DomainResponse>;
+  abstract getSubDomains(
+    id: string,
+    findOptions: Maybe<DomainFindOptions>
+  ): Promise<DomainsResponse>;
+  abstract searchDomainsByOwner(
+    address: string,
+    findOptions: Maybe<DomainFindOptions>
+  ): Promise<DomainsResponse>;
+  abstract searchDomainsByName(
+    searchTerm: string,
+    findOptions: Maybe<DomainFindOptions>
+  ): Promise<DomainsResponse>;
+  abstract doServiceOperation(serviceFn: Function): Promise<any>;
```

Comment:
`Promise<DomainsResponse | DomainResponse>`

Reply:
Forgot to type it, thanks